Filtered_kernel/include/CGAL/Lazy.h

```diff
@@ -230,59 +223,337 @@ struct Depth_base {
+ // Need some declarations, including Simple_cartesian.h would also be possible.
+ template<class>struct Simple_cartesian;
+ template<class>class Point_2;
+ template<class>class Point_3;
+template<bool b>struct Lazy_rep_selector<CGAL::Point_2<CGAL::Simple_cartesian<CGAL::Interval_nt<b>>>> { static constexpr int value = 1; };
+template<bool b>struct Lazy_rep_selector<CGAL::Point_3<CGAL::Simple_cartesian<CGAL::Interval_nt<b>>>> { static constexpr int value = 1; };
+# else
+template<bool b>struct Lazy_rep_selector<Interval_nt<b>> { static constexpr int value = 2; };
+# endif
+#else
+template<class AT>struct Lazy_rep_selector { static constexpr int value = 1; };
+#endif
+
+template<class AT>
+struct AT_wrap {
+  AT at_;
+  AT_wrap():at_(){}
+  AT_wrap(AT const& a):at_(a){}
+  AT_wrap(AT&& a):at_(std::move(a)){}
+  AT const& at()const{return at_;}
+};
+
+// TODO: avoid initializing AT for nothing
+template<class AT, class ET>
+struct AT_ET_wrap : AT_wrap<AT> {
+  ET et_;
+  AT_ET_wrap():et_(){}
+  AT_ET_wrap(ET const& e):et_(e){}
+  AT_ET_wrap(ET&& e):et_(std::move(e)){}
+  template<class A, class E>AT_ET_wrap(A&&a, E&&e):AT_wrap<AT>(std::forward<A>(a)),et_(std::forward<E>(e)){}
+  ET const& et()const{return et_;}
+};
 
 // Abstract base class for lazy numbers and lazy objects
-template <typename AT_, typename ET, typename E2A>
+template <typename AT_, typename ET, typename E2A, int=Lazy_rep_selector<AT_>::value /* 0 */>
 class Lazy_rep : public Rep, public Depth_base
 {
   Lazy_rep (const Lazy_rep&) = delete; // cannot be copied.
 
 public:
 
   typedef AT_ AT;
+  typedef AT_ET_wrap<AT,ET> Indirect;
 
-  mutable AT at;
-  mutable ET *et;
+  AT_wrap<AT> at_orig{};
+  mutable std::atomic<AT_wrap<AT>*> ptr_ { &at_orig };
+  mutable std::once_flag once;
 
-  Lazy_rep ()
-    : at(), et(nullptr){}
+  Lazy_rep () {}
 
   template<class A>
   Lazy_rep (A&& a)
-      : at(std::forward<A>(a)), et(nullptr){}
+      : at_orig(std::forward<A>(a)){}
 
   template<class A>
   Lazy_rep (int count, A&& a)
-    : Rep(count), at(std::forward<A>(a)), et(nullptr){}
+    : Rep(count), at_orig(std::forward<A>(a)){}
 
   template<class A, class E>
   Lazy_rep (A&& a, E&& e)
-      : at(std::forward<A>(a)), et(new ET(std::forward<E>(e))) {}
+      : ptr_(new AT_ET_wrap<AT,ET>(std::forward<A>(a), std::forward<E>(e))) {}
 
-  const AT& approx() const
+  AT const& approx() const
   {
-      return at;
+    return ptr_.load(std::memory_order_consume)->at();
   }
 
-  AT& approx()
+  const ET & exact() const
   {
-      return at;
+#if defined(CGAL_HAS_THREADS) && defined(__gnu_linux__) && __GNUC__ < 11 && !defined(_REENTRANT)
+    // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55394
+    // -pthread or -lpthread is just needed for linking, but I can't test that here...
+    // We could take the lack of _REENTRANT on that platform as forcing CGAL_HAS_NO_THREADS
+    static_assert(false, "Please pass -pthread to the compiler");
```

Comment:
Please make that assertion type-dependent, so that it is only evaluated at the instantiation of `Lazy_rep<AT,ET,E2A>::exact()`.

See for example https://devblogs.microsoft.com/oldnewthing/20200311-00/?p=103553.


We will have to work at the CMake level, to properly add `-pthread` when it is required.

Reply:
> Please make that assertion type-dependent,

That makes sense (for now the static_assert was just a note to myself so I wouldn't repeatedly hit mysterious errors), but I am not sure at all we want to keep this static_assert. We could undefine CGAL_HAS_THREADS on some platforms if it is missing, or some other approach.

Reply:
For a start, we should try a test platform without `CGAL_HAS_THREAD`, in CGAL test suite. I am not sure that still compiles everywhere.